py/emitinlinerv32.c

```diff
@@ -0,0 +1,774 @@
+                if ((kind & U) == 0) {
+                    if (!is_in_signed_mask(mask, immediate)) {
+                        goto out_of_range;
+                    }
+                } else {
+                    goto out_of_range;
+                }
+            }
+
+            if ((kind & Z) && (immediate == 0)) {
+                goto zero_immediate;
+            }
+
+            return true;
+        }
+        break;
+
+        case L: {
+            if (!MP_PARSE_NODE_IS_ID(node)) {
+                emit_inline_rv32_error_exc(emit,
+                    mp_obj_new_exception_msg_varg(&mp_type_SyntaxError,
+                        MP_ERROR_TEXT("opcode '%q' argument %d must be a label"),
+                        opcode_qstr, node_index + 1));
+                return false;
+            }
+
+            qstr qstring;
+            mp_uint_t label_index = lookup_label(emit, node, &qstring);
+            if (label_index >= emit->max_num_labels && emit->pass == MP_PASS_EMIT) {
+                emit_inline_rv32_error_exc(emit,
+                    mp_obj_new_exception_msg_varg(&mp_type_SyntaxError,
+                        MP_ERROR_TEXT("opcode '%q' argument %d label '%q' is undefined"),
+                        opcode_qstr, node_index + 1, qstring));
+                return false;
+            }
+
+            mp_uint_t displacement = (mp_uint_t)(label_code_offset(emit, label_index));
+            if (!is_in_unsigned_mask(mask, displacement)) {
+                if ((kind & U) == 0) {
+                    if (!is_in_signed_mask(mask, displacement)) {
+                        goto out_of_range;
+                    }
+                } else {
+                    goto out_of_range;
+                }
+            }
+            return true;
+        }
+        break;
+
+        default:
+            assert(!"Unknown argument kind");
+            break;
+    }
+
+    return false;
+
+out_of_range:
+    emit_inline_rv32_error_exc(emit,
+        mp_obj_new_exception_msg_varg(&mp_type_SyntaxError,
+            MP_ERROR_TEXT("opcode '%q' argument %d is out of range"),
+            opcode_qstr, node_index + 1));
+    return false;
+
+zero_immediate:
+    emit_inline_rv32_error_exc(emit,
+        mp_obj_new_exception_msg_varg(&mp_type_SyntaxError,
+            MP_ERROR_TEXT("opcode '%q' argument %d must not be zero"),
+            opcode_qstr, node_index + 1));
+    return false;
+}
+
+static bool parse_register_offset_node(emit_inline_asm_t *emit, qstr opcode_qstr, const opcode_t *opcode_data, mp_parse_node_t node, mp_uint_t node_index, mp_parse_node_t *register_node, mp_parse_node_t *offset_node, bool *negative) {
+    assert(register_node != NULL && "Register node pointer is NULL.");
+    assert(offset_node != NULL && "Offset node pointer is NULL.");
+    assert(negative != NULL && "Negative pointer is NULL.");
+
+    if (!MP_PARSE_NODE_IS_STRUCT_KIND(node, PN_atom_expr_normal) && !MP_PARSE_NODE_IS_STRUCT_KIND(node, PN_factor_2)) {
```

Comment:
This validation logic can be simplified, in pseudo code:
```
if MP_PARSE_NODE_IS_STRUCT_KIND(node, PN_factor_2):
    if MP_PARSE_NODE_IS_TOKEN_KIND(node_struct->nodes[0], MP_TOKEN_OP_MINUS):
        *negative = true;
    elif MP_PARSE_NODE_IS_TOKEN_KIND(node_struct->nodes[0], MP_TOKEN_OP_PLUS):
        // pass
    else:
        goto invalid;
    node = node_struct->nodes[1]

if !MP_PARSE_NODE_IS_STRUCT_KIND(node, PN_atom_expr_normal):
    goto invalid
```

Note that `PN_factor_2` and `PN_atom_expr_normal` always have 2 elements so there's no need to check that.

Reply:
Thanks, I ended up doing something similar, but I left the elements count check.  I'll take that out.

Right now I'm cleaning up the range checks, as negative offsets are still seen as positive due to the offset node being without its minus sign and you can imagine the rest.  Right now there are lots of repeated code for the special cases and I'm merging together.